### ISSUE TITLE
What is the difference between vector search and index search? [modular rag, chapter 1]

### ISSUE THREAD

**Issue by infinity-void6** [Denis2054/RAG-Driven-Generative-AI#2]
```
class RetrievalComponent(self,records,method='vector):
|
|
|
  def fit(self,records):
       if self.method=='vector' or self.method=='indexed':
         self.tfidf_matrix=self.vectorizer.fit_transform(records)
|
|
|
  def vector_search(self,query):
        query_tfidf=vectorizer.transform([query])
        similarities=cosine_similarity([query_tfidf,self.tfidf_matrix])
        best_index=similarities.argmax()
        return self.records[best_index]
  def indexed_search(self, query):
        # Assuming the tfidf_matrix is precomputed and stored
        query_tfidf = self.vectorizer.transform([query])
        similarities = cosine_similarity(query_tfidf, self.tfidf_matrix)
        best_index = similarities.argmax()
        return db_records[best_index]
```

What is the difference between vector search and indexed search? It seems to be the same. It is mentioned in the book that if tf_idf matrix is precomputed and stored, it is indexed search. But what does that actually mean to this code as we generate  tfidf_matrix for vector and index based search at fit()?

**Comment by infinity-void6** [comment]
based on what I got from chatgpt, vectors for vector search is gotten using something similar to sentence transformers and index based search uses Tf-idf Matrix for the arrays, am I right?

**Comment by infinity-void6** [comment]
won't vector search rather be something like 

```
from sentence_transformers import SentenceTransformer
model = SentenceTransformer("sentence-transformers/all-mpnet-base-v2")
#vector search
import numpy as np
def cosine_similarity(embedding1,embedding2):
      # Compute the dot product
    dot_product = np.dot(embedding1, embedding2)
    
    # Compute the norms
    norm1 = np.linalg.norm(embedding1)
    norm2 = np.linalg.norm(embedding2)
    
    # Compute cosine similarity
    if norm1 == 0 or norm2 == 0:  # Avoid division by zero
        return 0.0
    
    return dot_product / (norm1 * norm2)
def vector_search(query,dbrecords):
  embedding_query=model.encode(query)
  for record in dbrecords:
    record=record.lower()
    embedding_record=model.encode(record)
    similarity=cosine_similarity(embedding_query,embedding_record)
    best_index=similarity.argmax()
  return dbrecords[best_index]
```

**Comment by Denis2054** [comment]
The difference between vector and index search is:

**vector search**: Searches through each document and calculates the similarity.

Yes, the `def vector_search(self, query):` function could use the brute-force method implemented in `Section 3.1.Vector search` of this notebook.  In this case, the `def vector_search(self, query):` uses `tfidf_matrix`to increase the vector search performance.
Platforms do automatically provide the best functionality as implemented in  [Chapter 2 with vector embedding]( https://github.com/Denis2054/RAG-Driven-Generative-AI/blob/main/Chapter02/2_Embeddings_vector_store.ipynb)


**indexed search**: Uses a precomputed index (TF-IDF matrix) to compute cosine similarities.
The function for index search begins with:
`# Assuming the tfidf_matrix is precomputed and stored`, which platforms do automatically, as implemented in
[Chapter 3 for advanced indexing](https://github.com/Denis2054/RAG-Driven-Generative-AI/tree/main/Chapter03)
